File: SourceAFIS/Extraction/Templates/ITemplateFormat.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace SourceAFIS.Extraction.Templates
{
    public interface ITemplateFormat<T> where T : class
    {
        T Export(TemplateBuilder builder);
        TemplateBuilder Import(T template);
        void Serialize(Stream stream, T template);
        T Deserialize(Stream stream);
        
        byte[] Serialize(T template);
        T Deserialize(byte[] serialized);
        byte[] SerializeBuilder(TemplateBuilder builder);
        TemplateBuilder DeserializeBuilder(byte[] serialized);
    }
}

File: java/sourceafis/src/sourceafis/templates/ITemplateFormat.java
package sourceafis.templates;

import java.io.InputStream;
import java.io.OutputStream;

public interface ITemplateFormat<T>{
    T exportTemplate(TemplateBuilder builder);
    TemplateBuilder importTemplate(T template);
    void serialize(OutputStream stream, T template);
    T deserialize(InputStream stream);
    byte[] serialize(T template);
    T deserialize(byte[] serialized);
    byte[] serializeBuilder(TemplateBuilder builder);
    TemplateBuilder deserializeBuilder(byte[] serialized);
}